streams/shared/src/main/scala/zio/stream/ZTransducer.scala

```diff
@@ -120,6 +121,51 @@ abstract class ZTransducer[-R, +E, -I, +O](val push: ZManaged[R, Nothing, Option
       }
     }
 
+  /**
+   * Like [[zip]], but widens input type to `Either[I, I1]`.
+   */
+  def choose[R1 <: R, E1 >: E, I1, O1](
+    that: ZTransducer[R1, E1, I1, O1]
+  ): ZTransducer[R1, E1, Either[I, I1], Either[O, O1]] = {
+
+    //cross-compiling `Either.right`
+    @com.github.ghik.silencer.silent("deprecated")
+    @tailrec
+    def groupSides(
+      in: Chunk[Either[I, I1]],
+      acc: ChunkBuilder[Either[Some[Chunk[I]], Some[Chunk[I1]]]]
+    ): Chunk[Either[Some[Chunk[I]], Some[Chunk[I1]]]] = {
+      def impossible: Nothing = sys.error("impossible")
+      in.headOption match {
+        case Some(Left(_)) =>
+          val (init, rest) = in.span(_.isLeft)
+          groupSides(rest, acc += Left(Some(init.map(_.left.getOrElse(impossible)))))
+        case Some(Right(_)) =>
+          val (init, rest) = in.span(_.isRight)
+          groupSides(rest, acc += Right(Some(init.map(_.right.getOrElse(impossible)))))
+        case None => acc.result()
+      }
```

Comment:
Deconstructing chunks like this is highly inefficient. Let's convert this to a fold or a plain `while` loop.